### PR TITLE
add support for keep-alive timeout server hint

### PR DESCRIPTION
**Pull request by ronag**

### CODE REVIEW

**Comment by ronag** [comment]
@szmarczak

**Comment by ronag** [comment]
https://github.com/nodejs/node/issues/34560

File: lib/client.js [hunk 1]
```diff
@@ -435,12 +439,23 @@ class Parser extends HTTPParser {
...
 
+    if (headers['keep-alive']) {
+      const m = headers['keep-alive'].match(/timeout=(\d+)/)
+      if (m) {
+        const keepAliveTimeout = parseInt(m[1]) * 1000
+        // Set timeout to half of hint to account for timing inaccuracies.
+        client[kKeepAliveTimeout] = Math.min(keepAliveTimeout / 2, client[kIdleTimeout])
+      }
+    } else {
+      client[kKeepAliveTimeout] = client[kIdleTimeout]
```
Comment (szmarczak) [inline lib/client.js:450]: Is this line necessary?

File: lib/client.js [hunk 2]
```diff
@@ -435,12 +439,23 @@ class Parser extends HTTPParser {
       return true
     }
 
+    if (headers['keep-alive']) {
+      const m = headers['keep-alive'].match(/timeout=(\d+)/)
```
Comment (szmarczak) [inline lib/client.js:443]: Maybe let's define the regex at the top of the file, so we don't have to generate it every time?

File: lib/client.js [hunk 3]
```diff
@@ -435,12 +439,23 @@ class Parser extends HTTPParser {
...
+      if (m) {
+        const keepAliveTimeout = parseInt(m[1]) * 1000
+        // Set timeout to half of hint to account for timing inaccuracies.
+        client[kKeepAliveTimeout] = Math.min(keepAliveTimeout / 2, client[kIdleTimeout])
```
Comment (szmarczak) [inline lib/client.js:447]: `/ 2` seems to be too much. Why not `keepAliveTimeout - 100`? That should be pretty accurate even on slow machines.

File: lib/client.js [hunk 4]
```diff
@@ -435,12 +439,23 @@ class Parser extends HTTPParser {
       return true
     }
 
+    if (headers['keep-alive']) {
+      const m = headers['keep-alive'].match(/timeout=(\d+)/)
+      if (m) {
+        const keepAliveTimeout = parseInt(m[1]) * 1000
```
Comment (szmarczak) [inline lib/client.js:445]: What about `Number(m[1])`?

File: lib/client.js (hunk 2 above)
Comment (ronag) [reply lib/client.js:443]: > so we don't have to generate it every time?

I think v8 optimizes it out as a constant?

File: lib/client.js (hunk 1 above)
Comment (ronag) [reply lib/client.js:450]: I think so. If the server does not provide a hint for subsequent requests I think we should fallback to the idle timeout option.

File: lib/client.js (hunk 2 above)
Comment (szmarczak) [reply lib/client.js:443]: I have no deep idea about V8 optimizations.

The only thing I know is just that lambdas should be faster :P

File: lib/client.js (hunk 1 above)
Comment (szmarczak) [reply lib/client.js:450]: But you already set this at line 137, no?

File: lib/client.js (hunk 2 above)
Comment (ronag) [reply lib/client.js:443]: lambdas?

Comment (szmarczak) [reply lib/client.js:443]: In general. I don't know anything 'bout V8 optimizes variables.

File: lib/client.js (hunk 1 above)
Comment (ronag) [reply lib/client.js:450]: Not sure I follow. Line 447 will override line 137.

Comment (szmarczak) [reply lib/client.js:450]: I mean line 450 is the same as 137.

Comment (ronag) [reply lib/client.js:450]: no, 137 is in the constructor, the value might have been changed due to a response, however if a following response does not have a keep-alive hint it should go back.

Comment (szmarczak) [reply lib/client.js:450]: Ah, I get it now :D

**Pull request merged**